iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl

```diff
@@ -65,121 +92,156 @@ inline bool SpscSofi<ValueType, CapacityValue>::setCapacity(const uint64_t newSi
     uint64_t nextReadPosition{0};
 
-    bool popWasSuccessful{true};
     do
     {
+        // SYNC POINT READ: m_data
+        // See explanation of the corresponding synchronization point in push()
         if (currentReadPosition == m_writePosition.load(std::memory_order_acquire))
         {
             nextReadPosition = currentReadPosition;
-            popWasSuccessful = false;
+            return false;
+            // We don't need to check if read has changed, as it is enough to know that the empty
+            // state was valid in the past. The same race can also happen after the while loop and
+            // before the return operation
         }
         else
         {
-            // we use memcpy here, since the copy assignment is not thread safe in general (we might have an overflow in
-            // the push thread and invalidates the object while the copy is running and therefore works on an
-            // invalid object); memcpy is also not thread safe, but we discard the object anyway and read it
-            // again if its overwritten in between; this is only relevant for types larger than pointer size
-            // assign the user data
+            // we use memcpy here, to ensure that there is no logic in copying the data
             std::memcpy(&valueOut, &m_data[currentReadPosition % m_size], sizeof(ValueType));
-
-            /// @brief first we need to peak valueOut if it is fitting the condition and then we have to verify
-            ///        if valueOut is not am invalid object, this could be the case if the read position has
-            ///        changed
-            if (m_readPosition.load(std::memory_order_relaxed) == currentReadPosition && !verificator(valueOut))
-            {
-                popWasSuccessful = false;
-                nextReadPosition = currentReadPosition;
-            }
-            else
-            {
-                nextReadPosition = currentReadPosition + 1U;
-                popWasSuccessful = true;
-            }
         }
 
-        // compare and swap
-        // if(m_readPosition == currentReadPosition)
-        //     m_readPosition = l_next_aba_read_pos
-        // else
-        //     currentReadPosition = m_readPosition
-        // Assign m_aba_read_p to next readable location
+        // We need to check if m_readPosition hasn't changed otherwise valueOut might be corrupted
+        // While memory synchronization is not needed for m_readPosition, we need to ensure that the
+        // 'memcpy' happens before the CAS operation.
+        // Corresponding m_readPosition.load(acquire) is in the push method
+        // =============================================
+        // ABA problem: m_readPosition is an uint64_t. Assuming a thread is pushing at a rate of 1 GHz
+        // while this thread is blocked, we would still need more than 500 years to overflow
+        // m_readPosition and encounter the ABA problem
```

Comment:
It's not only the corruption of `valueOut` but we also need to prevent to reorder the memcpy after updating `m_readPosition`. With `relaxed` it would be perfectly fine for the compiler and CPU to successfully finish the while loop and then do the memcpy only once afterwards.

Reply:
Correct